sync15-adapter/src/request.rs

```diff
@@ -0,0 +1,1203 @@
+            assert!(!self.in_batch(),
+                    "Bug: Somehow we're in a batch but have no queued records");
+            // Nothing to do!
+            return Ok(());
+        }
+
+        self.queued.push(b']');
+        let batch_id = match &self.batch {
+            // Not the first post and we know we have no batch semantics.
+            &BatchState::Unsupported => None,
+            // First commit in possible batch
+            &BatchState::NoBatch => Some("true".into()),
+            // In a batch and we have a batch id.
+            &BatchState::InBatch(ref s) => Some(s.clone())
+        };
+
+        info!("Posting {} records of {} bytes", self.post_limits.cur_records, self.queued.len());
+
+        let is_commit = want_commit && !batch_id.is_none();
+        // Weird syntax for calling a function object that is a property.
+        let resp_or_error = self.poster.post(&self.queued,
+                                             self.last_modified,
+                                             batch_id,
+                                             is_commit,
+                                             self);
+
+        self.queued.truncate(0);
+
+        if want_commit || self.batch == BatchState::Unsupported {
+            self.batch_limits.clear();
+        }
+        self.post_limits.clear();
+
+        let resp = resp_or_error?;
+
+        if !resp.status.is_success() {
+            self.on_response.handle_response(resp, !want_commit)?;
+            bail!(error::unexpected("Expected OnResponse to have bailed out!"));
+        }
+
+        if want_commit {
+            debug!("Committed batch {:?}", self.batch);
+            self.batch = BatchState::NoBatch;
+            self.last_modified = resp.last_modified;
+            self.on_response.handle_response(resp, false)?;
+            return Ok(());
+        }
+
+        if resp.status != StatusCode::Accepted {
+            if self.in_batch() {
+                bail!(error::unexpected(
+                    "Server responded non-202 success code while a batch was in progress"));
+            }
+            self.last_modified = resp.last_modified;
+            self.batch = BatchState::Unsupported;
+            self.batch_limits.clear();
+            self.on_response.handle_response(resp, false)?;
+            return Ok(());
+        }
+
+        let batch_id = resp.result.batch.as_ref().ok_or_else(||
+            error::unexpected("Invalid server response: 202 without a batch ID"))?.clone();
+
+        match &self.batch {
+            &BatchState::Unsupported => {
+                warn!("Server changed it's mind about supporting batching mid-batch...");
+            },
+
+            &BatchState::InBatch(ref cur_id) => {
+                if cur_id != &batch_id {
+                    bail!(error::unexpected("Server changed batch id mid-batch!"));
+                }
+            },
+            _ => {}
+        }
+
+        // Can't change this in match arms without NLL
```

Comment:
What's "NLL"?

Reply:
I assume https://github.com/rust-lang/rust-roadmap/issues/16

Reply:
Yeah, there are a lot of cases where you can't modify a value because a reference is "alive" even though you aren't using it anymore. It's annoying.